### PR TITLE
Expand directories in CLI (2020)

### LINKED ISSUE #6085: support dot pattern for CLI
**Environments:**
- Prettier Version: 1.17.0
- Running Prettier via: CLI
- Runtime: Node.js
- Operating System: no matter

**Steps to reproduce:**

```shell
node_modules/.bin/prettier . --write
```

**Expected behavior:**

Format all supported files.

**Actual behavior:**

output:
```
[error] No matching files. Patterns tried: . !**/node_modules/** !./node_modules/** !**/.{git,svn,hg}/** !./.{git,svn,hg}/**
```

**Additional info:**

All popular linters (like `stylelint` and `eslint`) support `.` pattern. I think we should do same. I don't want have list of globs. It is good for DX and 0CJS configuration.

/cc @prettier/core What do you think? I can take care about this and send a PR.

### PR DESCRIPTION
**Pull request by thorn0**
Closes #6085
Supersedes #6128
Supersedes #6285

- [x] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by thorn0** [comment]
Resolved files aren't globally sorted anymore. Instead, their order matches the order of the given patterns, and for each pattern, the corresponding group of files is sorted. This seems logical and acceptable while ditching the global sorting allows us to use generators to avoid various problems related to big arrays.

Open problems:
1. files found by globs but all ignored by .prettierignore - warning?
2. directory expanding + --parser: what is "supported files" then?
3. directory expanding + glob-based config overrides: same question

File: src/cli/expand-patterns.js [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    };
+  }
+}
+
+const isWindows = path.sep === "\\";
+
+// TODO: use `fast-glob` directly, without `globby`
+const baseGlobbyOptions = {
+  dot: true,
+  expandDirectories: false,
```
Comment (fisker) [inline src/cli/expand-patterns.js]: Do we  support expand dirs in pattern? Like  `prettier **/foo` should work on any files in dir  named foo.  `expandDirectories` should work

File: src/cli/expand-patterns.js [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        context.languages.map(lang => lang.extensions || [])
+      );
+      const filenames = flat(
+        context.languages.map(lang => lang.filenames || [])
+      );
+      supportedFilesGlob = `**/{${extensions
+        .map(ext => "*" + (ext[0] === "." ? ext : "." + ext))
+        .concat(filenames)}}`;
+    }
+    return supportedFilesGlob;
```
Comment (fisker) [inline src/cli/expand-patterns.js]: Can I format  `foo.barjs` file with `config override `. I think right way is  glob all files
 and  `inferParser` then filter the files without  parser

File: tests_integration/__tests__/patterns-dirs.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    write: []
+  });
+});
+
+describe('4. `prettier . "dir2/**/*"`', () => {
+  runPrettier("cli/patterns-dirs", [".", "dir2/**/*", "-l"]).test({
+    status: 1,
+    write: []
+  });
+});
```
Comment (evilebottnawi) [inline tests_integration/__tests__/patterns-dirs.js]: Can we add tests with negative patterns like:
- `prettier dir !dir/nested`
- `prettier . !dir/nested`
- `prettier . !dir/nested/file.js`
- `prettier !dir !other` (should be error)

File: src/cli/expand-patterns.js [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    .relative(cwd, absolutePath)
+    .split(path.sep)
+    .some(dir => ignoredDirectories[dir]);
+}
+
+/**
+ * @param {string[]} paths
+ */
+function sortIfNotBig(paths) {
+  return paths.length > 10000
```
Comment (evilebottnawi) [inline src/cli/expand-patterns.js]: Can you clarify magic number?

File: src/cli/expand-patterns.js (hunk 2 above)
Comment (thorn0) [reply src/cli/expand-patterns.js]: > Can I format exts add by config override or plugins?

Plugins - yes. There is a test for that.

Config overrides - no. Because configs are resolved on a later stage.

> I think we should glob all files and inferParser then filter the files without parser

That's a good idea. I'll give it a try.

File: src/cli/expand-patterns.js (hunk 4 above)
Comment (thorn0) [reply src/cli/expand-patterns.js]: I'm not sure about this actually. It should probably be much bigger, like 1e7. What if Prettier is used to format millions of files? Wouldn't this sorting cause a useless slowdown?

Comment (thorn0) [reply src/cli/expand-patterns.js]: Thinking more about it, now I see that it's a premature optimization. I'll remove this condition. Let's fix it when we face it.

Comment (evilebottnawi) [reply src/cli/expand-patterns.js]: I think we always need to sort here, because we have to be deterministic, yes, it can be slow, for formatting really very very very big projects we should implement multithreading (maybe plus caching) (developers can write a custom script already).

Comment (thorn0) [reply src/cli/expand-patterns.js]: BTW, the sync API of `fast-glob` [returns results in a deterministic order](https://github.com/sindresorhus/globby/issues/131#issuecomment-590094878).

File: src/cli/expand-patterns.js (hunk 1 above)
Comment (thorn0) [reply src/cli/expand-patterns.js]: ESLint does this. Looks like we should too. But I doubt we can use globby's `expandDirectories` for that. Our expansion process is too custom, especially if we add parser inference to it.

Comment (thorn0) [reply src/cli/expand-patterns.js]: Sorry, I was wrong. ESLint actually **doesn't** support this. I forgot to put quotes around the glob, so it was resolved by Bash. So do we want to add this?

Comment (evilebottnawi) [reply src/cli/expand-patterns.js]: I am almost sure that we will receive a request for this in the near future with real-life examples, so will be great to support these patterns, if it does not require much effort, if it requires, we can postpone it

Comment (thorn0) [reply src/cli/expand-patterns.js]: This use case doesn't look very useful to me. What tools support this?

Comment (evilebottnawi) [reply src/cli/expand-patterns.js]: hm, eslint doesn't support, stylelint doesn't support, maybe @fisker known tools what supported that

File: tests_integration/__tests__/patterns-dirs.js (hunk 3 above)
Comment (thorn0) [reply tests_integration/__tests__/patterns-dirs.js]: > `prettier dir !dir/nested`

I don't think this should work. Negative patterns have to be globs (`!dir/nested/**/*`), not just directory paths with a preceding `!`. That's how glob resolution works.

Comment (evilebottnawi) [reply tests_integration/__tests__/patterns-dirs.js]: Why should not work? Get all files from `dir` and exclude all files from `dir/nested`

Comment (thorn0) [reply tests_integration/__tests__/patterns-dirs.js]: Okay, I was wrong. It should work, but I need to fix this first.

File: src/cli/expand-patterns.js (hunk 1 above)
Comment (fisker) [reply src/cli/expand-patterns.js]: If eslint and stylelint doesn't support, I'm ok we don't support this

File: src/cli/expand-patterns.js [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    } else {
+      entries.push({
+        type: "glob",
+        glob: fixWindowsSlashes(pattern),
+        input: pattern
+      });
+    }
+  }
+
+  for (const { type, absolutePath, glob, input } of entries) {
```
Comment (fisker) [inline src/cli/expand-patterns.js]: Do we need try dir first? For dir we ignore NoParserError, but for glob we print the error. What do we do with `prettier dir "dir/**/*"` and `prettier "dir/**/*" dir`? 

Still in hospital, can't play with it, just posting my concerts.

File: src/cli/expand-patterns.js [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+   *  glob?: string;
+   *  input: string;
+   * }>}
+   */
+  const entries = [];
+
+  for (const pattern of context.filePatterns) {
+    const absolutePath = path.resolve(cwd, pattern);
+
+    if (containsIgnoredPathSegment(absolutePath, cwd, ignoredDirectories)) {
```
Comment (evilebottnawi) [inline src/cli/expand-patterns.js]: What about using `micromatch` here, we already have it in our deps (from `fast-glob`)

File: src/cli/expand-patterns.js [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+const flat = require("lodash/flatten");
+
+/** @typedef {import('./util').Context} Context */
+
+/**
+ * @param {Context} context
+ */
+function* expandPatterns(context) {
+  const cwd = process.cwd();
+  const seen = Object.create(null);
```
Comment (evilebottnawi) [inline src/cli/expand-patterns.js]: Why do not use `Set`?

File: src/cli/expand-patterns.js [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+  for (const { type, absolutePath, glob, input } of entries) {
+    switch (type) {
+      case "file":
+        yield absolutePath;
+        continue;
+
+      case "dir": {
+        let result;
+        try {
+          result = globby.sync(glob, globbyOptions);
```
Comment (evilebottnawi) [inline src/cli/expand-patterns.js]: Looks here we have `expandDirectories: false`, so nested directories from `prettier dir` is not handled, right? I am on mobile, so can't test :smile:

File: src/cli/expand-patterns.js (hunk 5 above)
Comment (thorn0) [reply src/cli/expand-patterns.js]: I haven't implemented parser inference for dirs yet. But these are good test cases.

File: src/cli/expand-patterns.js (hunk 7 above)
Comment (thorn0) [reply src/cli/expand-patterns.js]: done

File: src/cli/expand-patterns.js (hunk 8 above)
Comment (thorn0) [reply src/cli/expand-patterns.js]: They're handled because a glob like this is composed for dirs: `<dir name>/**/<all supported extension and well-known filenames>`

When I implement [this suggestion](https://github.com/prettier/prettier/pull/7660#discussion_r383698135), it'll change to `<dir name>/**/*` with subsequent filtering based on parser inference.

File: src/cli/expand-patterns.js (hunk 6 above)
Comment (thorn0) [reply src/cli/expand-patterns.js]: This code is only for silent ignoring files in `node_modules` and VCS dirs. I think it's good enough.

File: tests_integration/__tests__/patterns-dirs.js (hunk 3 above)
Comment (thorn0) [reply tests_integration/__tests__/patterns-dirs.js]: Fixed it and added these tests.

File: src/cli/expand-patterns.js (hunk 5 above)
Comment (thorn0) [reply src/cli/expand-patterns.js]: > What do we do with `prettier dir "dir/**/*"` and `prettier "dir/**/*" dir`?

I think this is okay:
```console
$ ../../../bin/prettier.js dir1 'dir1/**/*' -l
dir1/a1.js
dir1/b1.js
dir1/nested1/an1.css
dir1/nested1/an1.js
dir1/nested1/bn1.js
dir1/a1.foo[error] No parser could be inferred for file: dir1/a1.foo
$ ../../../bin/prettier.js 'dir1/**/*' dir1 -l
dir1/a1.foo[error] No parser could be inferred for file: dir1/a1.foo
dir1/a1.js
dir1/b1.js
dir1/nested1/an1.css
dir1/nested1/an1.js
dir1/nested1/bn1.js
```

File: src/cli/expand-patterns.js (hunk 2 above)
Comment (thorn0) [reply src/cli/expand-patterns.js]: This parser inference thing looks a bit complicated though. How about merging this PR as is and improving this later?

I feel this PR is already good enough for most of the use cases. People just want to run `prettier . --write` or `prettier src --write`. For more complicated scenarios, they can use globs.

**Comment by fisker** [comment]
I guess only problem is how to get `supported files` in dir?

@thorn0

**Comment by fisker** [comment]
Can you also post a log run `./bin/prettier.js .` on codebase?

**Comment by thorn0** [comment]
> I guess only problem is how to get supported files in dir?

Yep, and the current solution is good enough, albeit not perfect.

**Comment by thorn0** [comment]
> Can you also post a log run `./bin/prettier.js .` on codebase?

With `--list-different`? Or what do you mean?

**Comment by thorn0** [comment]
<details>
<summary>
`./bin/prettier.js . --debug-check`
</summary>

```console
.azure-pipelines/jobs/dev-lint.yml
.azure-pipelines/jobs/dev-test.yml
.azure-pipelines/jobs/prod-build.yml
.azure-pipelines/jobs/prod-lint.yml
.azure-pipelines/jobs/prod-pack.yml
.azure-pipelines/jobs/prod-test.yml
.azure-pipelines/steps/download-dist.yml
.azure-pipelines/steps/install-dependencies.yml
.azure-pipelines/steps/install-nodejs.yml
.azure-pipelines/steps/publish-code-coverage.yml
.azure-pipelines/steps/publish-test-results.yml
.azure-pipelines/steps/upload-dist.yml
.codecov.yml
.eslintrc.yml
.github/FUNDING.yml
.github/ISSUE_TEMPLATE/formatting.md
.github/ISSUE_TEMPLATE/integration.md
.github/lock.yml
.github/no-response.yml
.github/PULL_REQUEST_TEMPLATE.md
.github/SECURITY.md
.github/workflows/dev-test.yml
.github/workflows/lint.yml
.github/workflows/prod-test.yml
.pre-commit-hooks.yaml
.prettierrc
.vscode/launch.json
.vscode/settings.json
azure-pipelines.yml
bin/prettier.js
changelog_unreleased/angular/pr-6869.md
changelog_unreleased/angular/pr-7371.md
changelog_unreleased/api/pr-5813.md
changelog_unreleased/api/pr-6908.md
changelog_unreleased/api/pr-6963.md
changelog_unreleased/api/pr-6993.md
changelog_unreleased/api/pr-7430.md
changelog_unreleased/api/pr-7435.md
changelog_unreleased/api/pr-7511.md
changelog_unreleased/api/pr-7620.md
changelog_unreleased/cli/pr-7210.md
changelog_unreleased/cli/pr-7587.md
changelog_unreleased/cli/pr-7588.md
changelog_unreleased/cli/pr-7660.md
changelog_unreleased/css/pr-6947.md
changelog_unreleased/flow/pr-7472.md
changelog_unreleased/flow/pr-7482.md
changelog_unreleased/flow/pr-7484.md
changelog_unreleased/graphql/pr-7305.md
changelog_unreleased/handlebars/pr-7051.md
changelog_unreleased/handlebars/pr-7052.md
changelog_unreleased/handlebars/pr-7157.md
changelog_unreleased/handlebars/pr-7275.md
changelog_unreleased/handlebars/pr-7306.md
changelog_unreleased/handlebars/pr-7552.md
changelog_unreleased/html/pr-5959.md
changelog_unreleased/html/pr-7273.md
changelog_unreleased/html/pr-7293.md
changelog_unreleased/html/pr-7295.md
changelog_unreleased/html/pr-7392.md
changelog_unreleased/html/pr-7395.md
changelog_unreleased/html/pr-7461.md
changelog_unreleased/html/pr-7555.md
changelog_unreleased/html/pr-7556.md
changelog_unreleased/html/pr-7654.md
changelog_unreleased/javascript/pr-3903.md
changelog_unreleased/javascript/pr-6685.md
changelog_unreleased/javascript/pr-6984.md
changelog_unreleased/javascript/pr-7010.md
changelog_unreleased/javascript/pr-7011.md
changelog_unreleased/javascript/pr-7026.md
changelog_unreleased/javascript/pr-7070.md
changelog_unreleased/javascript/pr-7087.md
changelog_unreleased/javascript/pr-7138.md
changelog_unreleased/javascript/pr-7140.md
changelog_unreleased/javascript/pr-7208.md
changelog_unreleased/javascript/pr-7367.md
changelog_unreleased/javascript/pr-7491.md
changelog_unreleased/javascript/pr-7500.md
changelog_unreleased/javascript/pr-7504.md
changelog_unreleased/markdown/pr-6852.md
changelog_unreleased/markdown/pr-7178.md
changelog_unreleased/markdown/pr-7181.md
changelog_unreleased/markdown/pr-7203.md
changelog_unreleased/mdx/pr-6398.md
changelog_unreleased/mdx/pr-6949.md
changelog_unreleased/scss/pr-6918.md
changelog_unreleased/scss/pr-7211.md
changelog_unreleased/TEMPLATE.md
changelog_unreleased/typescript/pr-6400.md
changelog_unreleased/typescript/pr-6901.md
changelog_unreleased/typescript/pr-7020.md
changelog_unreleased/typescript/pr-7075.md
changelog_unreleased/typescript/pr-7094.md
changelog_unreleased/typescript/pr-7104.md
changelog_unreleased/typescript/pr-7144.md
changelog_unreleased/typescript/pr-7221.md
changelog_unreleased/typescript/pr-7228.md
changelog_unreleased/vue/pr-7180.md
changelog_unreleased/vue/pr-7479.md
CHANGELOG.md
CODE_OF_CONDUCT.md
commands.md
CONTRIBUTING.md
docs/api.md
docs/browser.md
docs/cli.md
docs/comparison.md
docs/configuration.md
docs/editors.md
docs/for-enterprise.md
docs/ignore.md
docs/index.md
docs/install.md
docs/integrating-with-linters.md
docs/option-philosophy.md
docs/options.md
docs/plugins.md
docs/precommit.md
docs/rationale.md
docs/related-projects.md
docs/technical-details.md
docs/vim.md
docs/watching-files.md
docs/webstorm.md
docs/why-prettier.md
index.js
jest.config.js
package.json
README.md
scripts/build-docs.js
scripts/build/babel-plugins/transform-custom-require.js
scripts/build/build.js
scripts/build/bundler.js
scripts/build/cache.js
scripts/build/config.js
scripts/build/rollup-plugins/evaluate.js
scripts/build/rollup-plugins/executable.js
scripts/build/rollup-plugins/externals.js
scripts/build/rollup-plugins/native-shims.js
scripts/build/shims/assert.js
scripts/build/shims/events.js
scripts/build/shims/os.js
scripts/build/shims/path.js
scripts/build/shims/tty.js
scripts/build/util.js
scripts/check-deps.js
scripts/generate-schema.js
scripts/lint-changelog.js
scripts/release/package.json
scripts/release/README.md
scripts/release/release.js
scripts/release/steps/bump-prettier.js
scripts/release/steps/check-git-status.js
scripts/release/steps/generate-bundles.js
scripts/release/steps/install-dependencies.js
scripts/release/steps/post-publish-steps.js
scripts/release/steps/publish-to-npm.js
scripts/release/steps/push-to-git.js
scripts/release/steps/run-tests.js
scripts/release/steps/update-changelog.js
scripts/release/steps/update-version.js
scripts/release/steps/validate-new-version.js
scripts/release/utils.js
scripts/run-external-tests.js
scripts/sync-flow-tests.js
scripts/test-dist.js
src/cli/constant.js
src/cli/expand-patterns.js
src/cli/index.js
src/cli/minimist.js
src/cli/util.js
src/common/common-options.js
src/common/create-ignorer.js
src/common/end-of-line.js
src/common/errors.js
src/common/fast-path.js
src/common/get-file-info.js
src/common/internal-plugins.js
src/common/load-plugins.js
src/common/parser-create-error.js
src/common/parser-include-shebang.js
src/common/third-party.js
src/common/util-shared.js
src/common/util.js
src/config/resolve-config-editorconfig.js
src/config/resolve-config.js
src/document/doc-builders.js
src/document/doc-debug.js
src/document/doc-printer.js
src/document/doc-utils.js
src/document/index.js
src/index.js
src/language-css/clean.js
src/language-css/embed.js
src/language-css/index.js
src/language-css/options.js
src/language-css/parser-postcss.js
src/language-css/pragma.js
src/language-css/printer-postcss.js
src/language-css/utils.js
src/language-graphql/index.js
src/language-graphql/options.js
src/language-graphql/parser-graphql.js
src/language-graphql/pragma.js
src/language-graphql/printer-graphql.js
src/language-handlebars/clean.js
src/language-handlebars/index.js
src/language-handlebars/parser-glimmer.js
src/language-handlebars/printer-glimmer.js
src/language-handlebars/utils.js
src/language-html/ast.js
src/language-html/clean.js
src/language-html/conditional-comment.js
src/language-html/constants.evaluate.js
src/language-html/index.js
src/language-html/options.js
src/language-html/parser-html.js
src/language-html/pragma.js
src/language-html/preprocess.js
src/language-html/printer-html.js
src/language-html/syntax-attribute.js
src/language-html/syntax-vue.js
src/language-html/utils.js
src/language-js/clean.js
src/language-js/comments.js
src/language-js/embed.js
src/language-js/html-binding.js
src/language-js/index.js
src/language-js/loc.js
src/language-js/needs-parens.js
src/language-js/options.js
src/language-js/parser-angular.js
src/language-js/parser-babel.js
src/language-js/parser-flow.js
src/language-js/parser-typescript.js
src/language-js/postprocess.js
src/language-js/pragma.js
src/language-js/preprocess.js
src/language-js/printer-estree-json.js
src/language-js/printer-estree.js
src/language-js/utils.js
src/language-markdown/constants.evaluate.js
src/language-markdown/embed.js
src/language-markdown/index.js
src/language-markdown/mdx.js
src/language-markdown/options.js
src/language-markdown/parser-markdown.js
src/language-markdown/pragma.js
src/language-markdown/preprocess.js
src/language-markdown/printer-markdown.js
src/language-markdown/utils.js
src/language-yaml/index.js
src/language-yaml/options.js
src/language-yaml/parser-yaml.js
src/language-yaml/pragma.js
src/language-yaml/printer-yaml.js
src/language-yaml/utils.js
src/main/ast-to-doc.js
src/main/comments.js
src/main/core-options.js
src/main/core.js
src/main/massage-ast.js
src/main/multiparser.js
src/main/options-normalizer.js
src/main/options.js
src/main/parser.js
src/main/range-util.js
src/main/support.js
src/standalone.js
src/utils/arrayify.js
src/utils/create-language.js
src/utils/front-matter.js
src/utils/get-file-content-or-null.js
src/utils/get-last.js
src/utils/is-tty.js
src/utils/line-column-to-index.js
src/utils/load-toml.js
standalone.js
tests_config/fs.js
tests_config/require_prettier.js
tests_config/require_standalone.js
tests_config/run_spec.js
tests_integration/__tests__/arg-parsing.js
tests_integration/__tests__/check.js
tests_integration/__tests__/cli.js
tests_integration/__tests__/config-invalid.js
tests_integration/__tests__/config-resolution.js
tests_integration/__tests__/cursor-offset.js
tests_integration/__tests__/debug-check.js
tests_integration/__tests__/debug-print-doc.js
tests_integration/__tests__/doc-mark-as-root.js
tests_integration/__tests__/doc-trim.js
tests_integration/__tests__/early-exit.js
tests_integration/__tests__/file-info.js
tests_integration/__tests__/format.js
tests_integration/__tests__/help-options.js
tests_integration/__tests__/ignore-absolute-path.js
tests_integration/__tests__/ignore-in-subdirectories.js
tests_integration/__tests__/ignore-path.js
tests_integration/__tests__/ignore-relative-path.js
tests_integration/__tests__/ignore-vcs-files.js
tests_integration/__tests__/infer-parser.js
tests_integration/__tests__/invalid-ignore.js
tests_integration/__tests__/list-different.js
tests_integration/__tests__/load-toml.js
tests_integration/__tests__/loglevel.js
tests_integration/__tests__/parser-api.js
tests_integration/__tests__/patterns-dirs.js
tests_integration/__tests__/patterns-glob.js
tests_integration/__tests__/patterns.js
tests_integration/__tests__/piped-output.js
tests_integration/__tests__/plugin-default-options.js
tests_integration/__tests__/plugin-extensions.js
tests_integration/__tests__/plugin-options-string.js
tests_integration/__tests__/plugin-options.js
tests_integration/__tests__/plugin-precedence.js
tests_integration/__tests__/plugin-preprocess.js
tests_integration/__tests__/plugin-resolution.js
tests_integration/__tests__/plugin-virtual-directory.js
tests_integration/__tests__/schema.js
tests_integration/__tests__/skip-folders.js
tests_integration/__tests__/stdin-filepath.js
tests_integration/__tests__/support-info.js
tests_integration/__tests__/syntax-error.js
tests_integration/__tests__/util-shared.js
tests_integration/__tests__/with-config-precedence.js
tests_integration/__tests__/with-node-modules.js
tests_integration/__tests__/with-parser-inference.js
tests_integration/__tests__/with-shebang.js
tests_integration/__tests__/write.js
tests_integration/custom-parsers/custom-rename-input.js
tests_integration/custom-parsers/custom-rename-parser.js
tests_integration/env.js
tests_integration/path-serializer.js
tests_integration/runPrettier.js
tsconfig.json
website/.eslintrc.yml
website/blog/2017-04-13-1.0.0.md
website/blog/2017-04-20-1.2.0.md
website/blog/2017-05-03-1.3.0.md
website/blog/2017-06-03-1.4.0.md
website/blog/2017-06-28-1.5.0.md
website/blog/2017-08-29-1.6.0.md
website/blog/2017-09-15-1.7.0.md
website/blog/2017-11-07-1.8.0.md
website/blog/2017-12-05-1.9.0.md
website/blog/2018-01-10-1.10.0.md
website/blog/2018-02-26-1.11.0.md
website/blog/2018-04-09-plugin-php-0.1.md
website/blog/2018-04-11-1.12.0.md
website/blog/2018-05-27-1.13.0.md
website/blog/2018-07-29-1.14.0.md
website/blog/2018-11-07-1.15.0.md
website/blog/2019-01-20-1.16.0.md
website/blog/2019-04-12-1.17.0.md
website/blog/2019-06-06-1.18.0.md
website/blog/2019-11-09-1.19.0.md
website/core/Footer.js
website/data/editors.yml
website/data/languages.yml
website/data/users.yml
website/i18n/en.json
website/package.json
website/pages/en/help/index.js
website/pages/en/index.js
website/pages/en/users/index.js
website/pages/en/versions.js
website/pages/googlefe164a33bda4034b.html
website/pages/playground-redirect.html
website/pages/playground/index.html
website/playground/.eslintrc.yml
website/playground/BottomBar.js
website/playground/buttons.js
website/playground/codeSamples.js
website/playground/EditorState.js
website/playground/helpers.js
website/playground/index.js
website/playground/markdown.js
website/playground/panels.js
website/playground/Playground.js
website/playground/PrettierFormat.js
website/playground/sidebar/components.js
website/playground/sidebar/inputs.js
website/playground/sidebar/options.js
website/playground/sidebar/SidebarOptions.js
website/playground/storage.js
website/playground/urlHash.js
website/playground/util.js
website/playground/VersionLink.js
website/playground/WorkerApi.js
website/README.md
website/sidebars.json
website/siteConfig.js
website/static/install-service-worker.js
website/static/landing.js
website/static/manifest.json
website/static/overrides.css
website/static/playground.js
website/static/separate-css/playground.css
website/static/service-worker.js
website/static/worker.js
website/svgo.yml
website/versioned_docs/version-stable/api.md
website/versioned_docs/version-stable/browser.md
website/versioned_docs/version-stable/cli.md
website/versioned_docs/version-stable/comparison.md
website/versioned_docs/version-stable/configuration.md
website/versioned_docs/version-stable/editors.md
website/versioned_docs/version-stable/ignore.md
website/versioned_docs/version-stable/index.md
website/versioned_docs/version-stable/install.md
website/versioned_docs/version-stable/integrating-with-linters.md
website/versioned_docs/version-stable/option-philosophy.md
website/versioned_docs/version-stable/options.md
website/versioned_docs/version-stable/plugins.md
website/versioned_docs/version-stable/precommit.md
website/versioned_docs/version-stable/rationale.md
website/versioned_docs/version-stable/related-projects.md
website/versioned_docs/version-stable/technical-details.md
website/versioned_docs/version-stable/vim.md
website/versioned_docs/version-stable/watching-files.md
website/versioned_docs/version-stable/webstorm.md
website/versioned_docs/version-stable/why-prettier.md
website/versioned_sidebars/version-stable-sidebars.json
website/versions.json
website/webpack.config.js
```
</details>

**Comment by fisker** [comment]
> With `--list-different`? Or what do you mean?

Sorry, yes `./bin/prettier.js . --list-different`

**Comment by fisker** [comment]
I guess we can merge it first, and improve later, but if we are going to ship in 2.0 with current implementation, we should document the known issue(s).

**Comment by thorn0** [comment]
I documented it (cli.md) this way:

> If the path points to an existing directory, Prettier CLI recursively finds supported files in that directory. This resolution process is based on file extensions and well-known file names that Prettier and its [plugins](plugins.md) associate with supported languages.

**Comment by fisker** [comment]
I don't know if @lydell is available, but he might want review.

**Comment by fisker** [comment]
I forgot that when I was thinking about this feature, I'm going to do something like glob all files from dir and yield `{file, ignoreNoParserError: true}` , deal with it  later in `handleError`. This can handle unknown files add by config overrides.

It seems if we do this, the `NoResultsError` also need move there, I didn't think about this before.

**Comment by lydell** [comment]
What are the answers to the open problems?

> Open problems:
>
>  1.  files found by globs but all ignored by .prettierignore - warning?
>  2.  directory expanding + --parser: what is "supported files" then?
>  3.  directory expanding + glob-based config overrides: same question

**Comment by thorn0** [comment]
> What are the answers to the open problems?

# 2 and # 3 are the same problem, parser inference vs using extensions.

As for # 1, Prettier 1.19.1 doesn't issue any warning if passed a glob that is fully ignored by `.prettierignore`. Maybe it's not a big deal if the same will happen with directories.

**Comment by evilebottnawi** [comment]
@thorn0 

> directory expanding + --parser: what is "supported files" then?

Maybe we should using only `parser` extensions for better DX and 0cjs

**Comment by thorn0** [comment]
@evilebottnawi 

> > directory expanding + --parser: what is "supported files" then?

> Maybe we should using only parser extensions for better DX and 0cjs

I'd rather expect that `prettier dir --parser html` would match `dir/**/*` and use `html` for all files.

@fisker 

> glob all files from dir and yield `{file, ignoreNoParserError: true}` , deal with it later in handleError. This can handle unknown files add by config overrides. It seems if we do this, the `NoResultsError` also need move there, I didn't think about this before.

Yes. It's tricky. Not super difficult, but I'm sure that if I really start digging into it, I'll discover a dozen of new microissues there, so it might take long. The `.prettierignore` part looks especially suspicious. You tried to update the `ignore` package, so you should have an idea of what I mean.

**Comment by evilebottnawi** [comment]
> I'd rather expect that prettier dir --parser html would match dir/**/* and use html for all files.

Yes, i think you are right. Do we have a test for this?

**Comment by duailibe** [comment]
> I'd rather expect that prettier dir `--parser html` would match `dir/**/*` and use html for all files.

This should work exactly if we used `dir/**/*`, so you're right. It should just follow the current precedence order: command line args / config via API, config file (overrides and whatnot), inference based on extension.

IMO these things shouldn't block this PR if these problems already existed when using a blob, but those problems were things I wished to fix for a very long time (that would end up being a 2.0 release), that are way more important than other things people are more concerned with 😅

**Comment by thorn0** [comment]
>> I'd rather expect that prettier dir --parser html would match dir/**/* and use html for all files.

> Yes, i think you are right. Do we have a test for this?

That's not how it works now. It's the [same issue](https://github.com/prettier/prettier/pull/7660#discussion_r383698135), namely: what exactly is "supported files" when we're looking for all "supported files" in a directory?

The current implementation in this PR: "supported files" = files with extensions and names taken from `getSupportInfo()`.

An ideal implementation: "supported files" = `glob(dir + '/**').filter(parserCanBeInferred)`. In the case when `--parser` is specified as a CLI arg, `parserCanBeInferred` turns into `() => true`.

**Comment by duailibe** [comment]
Ah, now I understand the problem

> The current implementation in this PR: "supported files" = files with extensions and names taken from `getSupportInfo()`.

This sounds good to me

**Comment by thorn0** [comment]
My proposal is to postpone the ideal implementation until 3.0.

**Comment by evilebottnawi** [comment]
> My proposal is to postpone the ideal implementation until 3.0.

Sounds good for me, we will also collect feedback and it will help us make it better

**Comment by fisker** [comment]
Looks like @duailibe and @evilebottnawi both Okay with current implementation.

But I have a question, can we somehow hide this `NoParserError` for both dir and glob? I'm on cellphone, I don't know how `ESLint` handle `eslint *.jpg`? I guess it throw a `ParseError` and abort cli?

**Comment by thorn0** [comment]
@fisker Currently (in 1.19.1 and in this PR), the "parser can't be inferred" error is not fatal. The error is logged, the file is skipped, Prettier CLI proceeds with next file. Why would we want to hide the error? ESLint doesn't have relevant behavior because it always uses the same parser.

**Comment by fisker** [comment]
@thorn0 I was thinking if eslint abort linting when error occurs, we can also ask people pass glob with valid extensions like `prettier *.{js,css}` instead of `prettier *`.  For ESLint I never use `--ext`, I use [`**/*.{js,mjs}`](https://github.com/fisker/import-esm/blob/d1d16964f6fecb4388cedc03f3ff8141b876438c/package.json#L44)

**Comment by thorn0** [comment]
@fisker Not sure how what you wrote this time is different from what you wrote before about 'I was going to do something like glob all files from dir and yield {file, ignoreNoParserError: true}'.

**Comment by fisker** [comment]
I'm okay with current implementation.

This is only a idea of a new possible/simple solution.

**Comment by thorn0** [comment]
I'm actually trying to rewrite it. Let's see how it goes. If it works, I'll open a new PR.

**Comment by fisker** [comment]
Should we merge this and release v2.0?

**Comment by thorn0** [comment]
@evilebottnawi Okay, I'll do this today (if nobody objects).
